design/design-generic-references.md

```diff
@@ -0,0 +1,477 @@
+field containing the value used later for setting `spec.foo` is fixed, i.e. it
+is embedded in the resolving mechanism used by the provider.
+
+Although the used convention covers the majority of use cases, user experience
+shows that there is still a need to refer a value on an arbitrary path within an
+object of arbitrary kind. For example, 
+[`datasync.aws.upbound.io` `Task`](https://github.com/upbound/provider-aws/blob/main/apis/datasync/v1beta1/zz_task_types.go#L375)
+has [`spec.forProvider.sourceLocationArn`](https://github.com/upbound/provider-aws/blob/main/apis/datasync/v1beta1/zz_task_types.go#L359)
+field that can obtain value from an instance of the following object types:
+
+* LocationS3
+* LocationSmb
+* LocationObjectStorage
+* LocationNfs
+* LocationHdfs
+* LocationFsxWindowsFileSystem
+* LocationFsxOpenzfsFileSytem
+* LocationFsxLustreFileSystem
+* LocationEfs
+
+However, [the implemented provider resolving mechanism](https://github.com/upbound/provider-aws/blob/main/apis/datasync/v1beta1/zz_generated.resolvers.go#L104-L112)
+looks only for the value in `LocationS3` instance named
+by `spec.forProvider.sourceLocationArnRef` field.
+
+Currently, the issue can be solved by defining the following composition:
+
+* Declare under managed resources the needed source object (one of `Location*`)
+  types with [observe only management policy]
+* Patch composite object with the value from the source object
+* Patch Task's `spec.forProvider.sourceLocationArn` field with the composite
+  value
+
+The proposed solution has the following drawbacks:
+
+* For each claim, additional observe only resource gets created, just to be able
+  to work around the limitations of the current reference mechanism on certain
+  fields. In a system with a high number of claims, that would put additional
+  pressure on k8s API and etcd.
+* Composite object needs to expose the patched value, although fully unneeded
+  from API design perspective
+
+__**NOTE**__: [a prior design exists on this topic](https://github.com/crossplane/crossplane/pull/2385).
+
+## Goals
+
+The generic cross-reference should be able to:
+
+* Refer a value on an arbitrary path within an object of arbitrary kind
+* Use the existing [Kubernetes API object reference] convention as much as
+  possible
+* Be implementable without or requiring only light changes in RBAC
+
+In cases when the generic cross-references are unavailable for a given managed
+resource field, our compositions should not require declaring observe-only
+resource, for sole purpose of referring a value.
+
+## Proposal
+
+Until now, providers were supporting only reference fields of
+type [`Reference`](https://github.com/pedjak/crossplane-runtime/blob/master/apis/common/v1/resource.go#L116-L123):
+
+```go
+type Reference struct {
+// Name of the referenced object.
+Name string `json:"name"`
+
+// Policies for referencing.
+// +optional
+Policy *Policy `json:"policy,omitempty"`
```

Comment:
Out of scope for this design, but FYI the presence of this field is a bit of a problem - see https://github.com/crossplane/crossplane-runtime/issues/440.